# Rolex Listing Price Prediction based on model and complications

In [1]:
import pandas as pd
import numpy as np
import glob
import janitor
import altair as alt
alt.data_transformers.enable("vegafusion")

DataTransformerRegistry.enable('vegafusion')

## Data Cleaning

In [2]:
files = glob.glob('data/result_df/*.csv')
dirty_df = pd.concat((pd.read_csv(file, index_col=0)
                for file in files)
              )

dirty_df = dirty_df.clean_names()
dirty_df.drop_duplicates(subset=['listing_code', 'reference_number'], inplace=True)
dirty_df.dropna(subset=['brand', 'model', 'listing_code', 'price', 'title', 'subtitle', 'case_diameter'], inplace=True)
dirty_df.reset_index(drop=True, inplace=True)


dirty_df.head()

listing_code  brand          model  \
0       IJD7R3  Rolex    Datejust 41   
1       HOAOQ5  Rolex    Datejust 31   
2       IJ9RY8  Rolex    Datejust 36   
3       FDHJM3  Rolex  GMT-Master II   
4       FFF9D3  Rolex       Explorer   

                                reference_number   movement case_material  \
0  126331 NEW UNWORN 2023 Wimbledon 41mm Jubilee  Automatic    Gold/Steel   
1                                         278271  Automatic    Gold/Steel   
2                                         126231  Automatic    Gold/Steel   
3                                     126710BLNR  Automatic         Steel   
4                                         124270  Automatic         Steel   

  bracelet_material year_of_production  \
0        Gold/Steel               2023   
1        Gold/Steel               2023   
2        Gold/Steel               2023   
3             Steel               2023   
4             Steel               2021   

                                           condition  \
0        New\n(Brand new, without any signs of wear)   
1        New\n(Brand new, without any signs of wear)   
2        New\n(Brand new, without any signs of wear)   
3        New\n(Brand new, without any signs of wear)   
4  Very good\n(Worn with little to no signs of wear)   

               scope_of_delivery  ... thickness lug_width buckle_width  \
0  Original box, original papers  ...       NaN       NaN          NaN   
1  Original box, original papers  ...       NaN       NaN          NaN   
2  Original box, original papers  ...       NaN       NaN          NaN   
3  Original box, original papers  ...       NaN       NaN          NaN   
4  Original box, original papers  ...       NaN       NaN          NaN   

  frequency bracelet_thickness submariner_kermit_ref_ day_date_ref_  \
0       NaN                NaN                    NaN           NaN   
1       NaN                NaN                    NaN           NaN   
2       NaN                NaN                    NaN           NaN   
3       NaN                NaN                    NaN           NaN   
4       NaN                NaN                    NaN           NaN   

   datejust_reference_number submariner_date_reference reference  
0                        NaN                       NaN       NaN  
1                        NaN                       NaN       NaN  
2                        NaN                       NaN       NaN  
3                        NaN                       NaN       NaN  
4                        NaN                       NaN       NaN  

[5 rows x 48 columns]

In [3]:
# clean case_diameter
def is_convertible_to_int(value):
    try:
        int(value)
        return True
    except ValueError:
        return False

convertible_mask = dirty_df['case_diameter'].str[:2].apply(is_convertible_to_int)

dirty_df = dirty_df[convertible_mask]

dirty_df['case_diameter'] = dirty_df['case_diameter'].str[:2].astype('int')


In [4]:
# add column of whether the price is negotiable
dirty_df.insert(loc=13, column='is_negotiable', value=dirty_df['price'].str.contains('Negotiable', case=False).astype(int))

In [5]:
# keep only CA$ in the `price` column
dirty_df['price'] = dirty_df['price'].str.extract('C\$([0-9,]+)')[0].str.replace(',', '')
dirty_df['price'] = pd.to_numeric(dirty_df['price'], errors='coerce')
dirty_df['price'].fillna(0, inplace=True)
dirty_df['price'] = dirty_df['price'].astype(int)

dirty_df = dirty_df.query('price != 0')

<>:2: DeprecationWarning: invalid escape sequence '\$'
<>:2: DeprecationWarning: invalid escape sequence '\$'
C:\Users\fohy2\AppData\Local\Temp\ipykernel_6332\2926559523.py:2: DeprecationWarning: invalid escape sequence '\$'
  dirty_df['price'] = dirty_df['price'].str.extract('C\$([0-9,]+)')[0].str.replace(',', '')


In [6]:
# add column of whether the year of production is approximated
dirty_df.insert(loc=8, column='year_is_approximated', value=dirty_df['year_of_production'].str.contains('Approximation', case=False).astype(int))

# Clean year of production
dirty_df['year_of_production'] = dirty_df['year_of_production'].apply(lambda x: x[:4] if x != 'Unknown' else x)

In [7]:
# simplify the location to country only
dirty_df['country'] = dirty_df['location'].str.split(',').str[0]

Save the cleaned data locally

In [8]:
rolex_df = dirty_df
rolex_df.to_csv('data/rolex_df.csv')

## EDA

In [9]:
from sklearn.model_selection import cross_val_score, cross_validate, train_test_split

In [10]:
display(rolex_df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 62495 entries, 0 to 66279
Data columns (total 51 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   listing_code                          62495 non-null  object 
 1   brand                                 62495 non-null  object 
 2   model                                 62495 non-null  object 
 3   reference_number                      61846 non-null  object 
 4   movement                              61443 non-null  object 
 5   case_material                         60862 non-null  object 
 6   bracelet_material                     56783 non-null  object 
 7   year_of_production                    62495 non-null  object 
 8   year_is_approximated                  62495 non-null  int32  
 9   condition                             61537 non-null  object 
 10  scope_of_delivery                     62495 non-null  object 
 11  gender              

None

In [11]:
df = rolex_df[['model', 'movement', 'case_material', 'bracelet_material',
               'year_of_production', 'year_is_approximated', 'condition', 'scope_of_delivery',
               'country', 'availability', 'case_diameter', 'bezel_material',
               'crystal', 'dial', 'bracelet_color', 'clasp', 'clasp_material',
               'rating', 'reviews', 'price', 'is_negotiable']]
df.head(1)

model   movement case_material bracelet_material year_of_production  \
0  Datejust 41  Automatic    Gold/Steel        Gold/Steel               2023   

   year_is_approximated                                    condition  \
0                     0  New\n(Brand new, without any signs of wear)   

               scope_of_delivery                   country      availability  \
0  Original box, original papers  United States of America  Item is in stock   

   ...  bezel_material           crystal    dial bracelet_color       clasp  \
0  ...       Rose gold  Sapphire crystal  Silver     Gold/Steel  Fold clasp   

  clasp_material rating  reviews  price  is_negotiable  
0     Gold/Steel    4.2       11  23421              1  

[1 rows x 21 columns]

In [12]:
df.shape

(62495, 21)

We will use only the following columns since they have fewer missing values and have more variation even for the same model. Features that are unrelated to the watch model is especially interesting, such as `condition` and `scope_of_delivery`, as they provide insights on how these factor in to the listing price.

In [13]:
train_df, test_df = train_test_split(df, test_size=0.3, random_state=123)
print(train_df.shape)
print(test_df.shape)

(43746, 21)
(18749, 21)


In [14]:
X_train, y_train = train_df.drop(
    columns=["price"]), train_df["price"]
y_train = pd.DataFrame(y_train)
X_test, y_test = test_df.drop(
    columns=["price"]), test_df["price"]
y_test = pd.DataFrame(y_test)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(43746, 20)
(43746, 1)
(18749, 20)
(18749, 1)


In [15]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43746 entries, 63205 to 56465
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 43746 non-null  object 
 1   movement              43021 non-null  object 
 2   case_material         42626 non-null  object 
 3   bracelet_material     39833 non-null  object 
 4   year_of_production    43746 non-null  object 
 5   year_is_approximated  43746 non-null  int32  
 6   condition             43093 non-null  object 
 7   scope_of_delivery     43746 non-null  object 
 8   country               43746 non-null  object 
 9   availability          43746 non-null  object 
 10  case_diameter         43746 non-null  int32  
 11  bezel_material        32235 non-null  object 
 12  crystal               35703 non-null  object 
 13  dial                  40357 non-null  object 
 14  bracelet_color        33746 non-null  object 
 15  clasp               

In [16]:
plot_columns = X_train.columns.to_list()

for column in plot_columns:
    top_categories = X_train[column].value_counts().head(10).index
    filtered_X_train = X_train[X_train[column].isin(top_categories)]

    chart = alt.Chart(filtered_X_train).mark_bar().encode(
        y=alt.Y(f"{column}:N", sort='-x'),
        x=alt.X('count()', title='Count')
    ).properties(
        title=f"Top 10 Categories in {column}"
    )
    
    chart.display()

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

alt.Chart(...)

In [17]:
y_train.describe(percentiles=[.25, .5, .75, 0.975]).apply(lambda s: s.apply('{0:.0f}'.format))

price
count    43746
mean     31721
std      43769
min        198
25%      13105
50%      20724
75%      33910
97.5%   119718
max    1506426

In [18]:
alt.Chart(y_train.query('price <= 120000'),
          title='Histogram of Rolex price').mark_bar().encode(
    alt.X('price:Q').bin(maxbins=40),
    y='count()'
)

alt.Chart(...)

The above histogram is showing at least 97.5% of the price data. It is difficult to interpret the distribution with the outliers so they are disregarded for the purpose of this visualization.

In [19]:
train_df.corr(numeric_only=True).round(
    decimals=3).style.background_gradient()

The price seems to be slightly positively correlated with case diameter, which is expected as larger models are usually equipped with more complications that drive up the price.

## Models

In [20]:
# imports
import sys, os
import time

import matplotlib.pyplot as plt

%matplotlib inline
import numpy as np
import pandas as pd
import altair as alt
from IPython.display import HTML

sys.path.append(os.path.join(os.path.abspath("."), "code"))

from IPython.display import display

# Classifiers and regressors
from sklearn.dummy import DummyClassifier, DummyRegressor

# Preprocessing and pipeline
from sklearn.impute import SimpleImputer

# train test split and cross validation
from sklearn.model_selection import cross_val_score, cross_validate, train_test_split
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.compose import ColumnTransformer, make_column_transformer
from sklearn.preprocessing import (
    MinMaxScaler,
    OneHotEncoder,
    OrdinalEncoder,
    StandardScaler,
)
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

from sklearn.feature_selection import SelectFromModel
from sklearn.linear_model import *
from sklearn.tree import *
from sklearn.ensemble import *
from lightgbm.sklearn import *

In [21]:
# adapted from 571 lecture notes
# code from lecture
def mean_std_cross_val_scores(model, X_train, y_train, **kwargs):
    """
    Returns mean and std of cross validation

    Parameters
    ----------
    model :
        scikit-learn model
    X_train : numpy array or pandas DataFrame
        X in the training data
    y_train :
        y in the training data

    Returns
    ----------
        pandas Series with mean scores from cross_validation
    """

    scores = cross_validate(model, X_train, y_train, n_jobs=-1, **kwargs)

    mean_scores = pd.DataFrame(scores).mean()
    std_scores = pd.DataFrame(scores).std()
    out_col = []

    for i in range(len(mean_scores)):
        out_col.append((f"%0.3f (+/- %0.3f)" %
                       (mean_scores.iloc[i], std_scores.iloc[i])))

    return pd.Series(data=out_col, index=mean_scores.index)

In [22]:
# One-hot encoding for categorical features

categorcial_feats = [col for col in X_train.columns if col not in ['case_diameter', 'rating', 'reviews']]
categorical_pipe = make_pipeline(OneHotEncoder(drop='if_binary', handle_unknown='ignore'))

preprocessor = make_column_transformer((categorical_pipe, categorcial_feats))
preprocessor

ColumnTransformer(transformers=[('pipeline',
                                 Pipeline(steps=[('onehotencoder',
                                                  OneHotEncoder(drop='if_binary',
                                                                handle_unknown='ignore'))]),
                                 ['model', 'movement', 'case_material',
                                  'bracelet_material', 'year_of_production',
                                  'year_is_approximated', 'condition',
                                  'scope_of_delivery', 'country',
                                  'availability', 'bezel_material', 'crystal',
                                  'dial', 'bracelet_color', 'clasp',
                                  'clasp_material', 'is_negotiable'])])

In [23]:
# create a dictionary for storing model scores
results_dict = {}

In [24]:
linear_reg = make_pipeline(preprocessor,
                           LinearRegression())
results_dict["linear regression"] = mean_std_cross_val_scores(
    linear_reg, X_train, y_train, return_train_score=True
)
pd.DataFrame(results_dict).T

fit_time         score_time         test_score  \
linear regression  1.287 (+/- 0.042)  0.120 (+/- 0.006)  0.437 (+/- 0.041)   

                         train_score  
linear regression  0.446 (+/- 0.011)

In [25]:
dt = make_pipeline(preprocessor, DecisionTreeRegressor())
results_dict["decision tree"] = mean_std_cross_val_scores(
    dt, X_train, y_train, return_train_score=True
)
pd.DataFrame(results_dict).T

fit_time         score_time         test_score  \
linear regression  1.287 (+/- 0.042)  0.120 (+/- 0.006)  0.437 (+/- 0.041)   
decision tree      7.587 (+/- 0.102)  0.089 (+/- 0.007)  0.187 (+/- 0.284)   

                         train_score  
linear regression  0.446 (+/- 0.011)  
decision tree      0.993 (+/- 0.001)

In [26]:
rf = make_pipeline(preprocessor, RandomForestRegressor(random_state=123))
results_dict["random forest"] = mean_std_cross_val_scores(
    rf, X_train, y_train, return_train_score=True
)
pd.DataFrame(results_dict).T

fit_time         score_time         test_score  \
linear regression    1.287 (+/- 0.042)  0.120 (+/- 0.006)  0.437 (+/- 0.041)   
decision tree        7.587 (+/- 0.102)  0.089 (+/- 0.007)  0.187 (+/- 0.284)   
random forest      508.343 (+/- 1.875)  0.385 (+/- 0.071)  0.578 (+/- 0.074)   

                         train_score  
linear regression  0.446 (+/- 0.011)  
decision tree      0.993 (+/- 0.001)  
random forest      0.940 (+/- 0.002)

In [ ]:
lightgbm = make_pipeline(preprocessor, LGBMRegressor(random_state=123))
results_dict["lightgbm"] = mean_std_cross_val_scores(
    lightgbm, X_train, y_train, return_train_score=True
)
pd.DataFrame(results_dict).T